File: shared/scripts/flux/_loader.ts
/**
 * flux spec loader — Node-side bridge invoked by the Python codegen pipeline.
 *
 * Usage (internal — called by ``shared/scripts/flux/loader.py``)::
 *
 *     npx tsx shared/scripts/flux/_loader.ts <spec-file>
 *
 * The script dynamic-imports the given ``.spec.ts`` file, grabs
 * ``mod.default`` (which ``defineComponent`` has already validated), then
 * walks the spec to produce a JSON-serialisable snapshot:
 *
 *   * ``schema`` (the Zod object) is converted via ``z.toJSONSchema`` and
 *     re-keyed as ``jsonSchema`` so the emitted catalog matches the shape
 *     required by ``catalog/flux/schema/component.schema.json``.
 *   * Everything else is copied verbatim in the key order the DSL author
 *     wrote — the loader does NOT reshape or normalise ordering so
 *     downstream emitters stay deterministic.
 *
 * The script prints exactly ONE JSON object (pretty-printed with 2-space
 * indent) to stdout and nothing else; every diagnostic goes to stderr so
 * the parent Python process can round-trip stdout → ``json.loads`` without
 * scrubbing.
 *
 * Exit codes::
 *
 *     0 — JSON printed to stdout
 *     1 — usage / import / validation error (details on stderr)
 */

import { pathToFileURL } from "node:url";
import { z } from "zod";

type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [key: string]: JsonValue };

type ZodLike = z.ZodType<unknown>;

function isZod(value: unknown): value is ZodLike {
  return (
    typeof value === "object" &&
    value !== null &&
    "_def" in (value as Record<string, unknown>) &&
    typeof (value as { parse?: unknown }).parse === "function"
  );
}

/**
 * Compile a Zod schema to JSON Schema via Zod v4's built-in converter. We
 * strip the top-level ``$schema`` key so the emitted catalog's component
 * block stays a plain JSON Schema object (the outer catalog carries its own
 * ``$schema`` reference).
 */
function toJsonSchema(schema: ZodLike): JsonValue {
  // `z.toJSONSchema` is the Zod v4 entry point.
  const out = z.toJSONSchema(schema);
  if (typeof out === "object" && out !== null && "$schema" in out) {
    const clone: Record<string, JsonValue> = { ...(out as Record<string, JsonValue>) };
    delete clone["$schema"];
    return clone;
  }
  return out as JsonValue;
}

/**
 * Walk the spec tree recursively, converting Zod schemas (found anywhere)
 * to JSON Schema and leaving every other value intact. The DSL only stores
 * one Zod instance (the ``schema`` field) but we walk defensively so that a
 * future spec that embeds Zod deeper still serialises correctly.
 */
function serialise(value: unknown): JsonValue {
  if (value === null || value === undefined) return null;
  if (isZod(value)) return toJsonSchema(value);
  if (Array.isArray(value)) return value.map(serialise);
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const out: Record<string, JsonValue> = {};
    for (const key of Object.keys(obj)) {
      const child = serialise(obj[key]);
      if (child !== undefined) out[key] = child;
    }
    return out;
  }
  if (typeof value === "function") {
    // Functions in a spec are a DSL-level mistake; surface them rather than
    // silently dropping.
    throw new Error(`flux loader: cannot serialise function value (${(value as Function).name || "<anon>"})`);
  }
  if (typeof value === "bigint") return value.toString();
  return value as JsonValue;
}

async function main(): Promise<void> {
  const specPath = process.argv[2];
  if (!specPath) {
    console.error("usage: _loader.ts <spec-path>");
    process.exit(1);
  }
  const url = pathToFileURL(specPath).href;
  let mod: { default?: unknown };
  try {
    mod = (await import(url)) as { default?: unknown };
  } catch (err) {
    console.error(`flux loader: failed to import ${specPath}`);
    console.error(err instanceof Error ? err.stack ?? err.message : String(err));
    process.exit(1);
    return;
  }
  if (!mod.default) {
    console.error(`flux loader: ${specPath} has no default export (expected ComponentSpec)`);
    process.exit(1);
    return;
  }
  // DSL-author spec — shape comes from dsl/types.ts#ComponentSpec. We pass
  // the whole thing through ``serialise`` which swaps the Zod schema for
  // JSON Schema and renames the key.
  const spec = mod.default as Record<string, unknown>;
  const serialised = serialise(spec) as Record<string, JsonValue>;
  // Rename `schema` → `jsonSchema` to match component.schema.json.
  if ("schema" in serialised) {
    serialised.jsonSchema = serialised.schema;
    delete serialised.schema;
  }
  // Deterministic key order: rebuild the object in the canonical order
  // described by ARCHITECTURE.md §6 + the component schema. Any future key
  // that ARCHITECTURE adds lands alphabetically at the end so emitters stay
  // stable.
  const CANONICAL_ORDER = [
    "name",
    "extends",
    "category",
    "jsonSchema",
    "slots",
    "tokens",
    "variants",
    "compoundVariants",
    "defaultVariants",
    "accessibility",
    "llm",
    "renderer",
  ];
  const ordered: Record<string, JsonValue> = {};
  for (const key of CANONICAL_ORDER) {
    if (key in serialised) ordered[key] = serialised[key];
  }
  for (const key of Object.keys(serialised).sort()) {
    if (!(key in ordered)) ordered[key] = serialised[key];
  }
  process.stdout.write(JSON.stringify(ordered, null, 2) + "\n");
}

main().catch((err) => {
  console.error("flux loader: uncaught error");
  console.error(err instanceof Error ? err.stack ?? err.message : String(err));
  process.exit(1);
});

File: shared/scripts/flux/loader.py
"""Run the TypeScript loader on every ``specs/*.spec.ts`` and return a dict.

The loader shells out to ``npx tsx shared/scripts/flux/_loader.ts`` with
``cwd=ts/`` and ``NODE_PATH=ts/node_modules`` so Zod + zod-to-json-schema
resolve via the TS workspace's ``package.json``. The spec file itself
lives under ``catalog/flux/specs/``; we pass it as an absolute path so
``pathToFileURL`` in the Node loader converts correctly.

Return shape::

    {
        "FluxButton": { ...component dict... },
        "FluxCard":   { ... },
        ...
    }

Keys are the ``name`` field of each spec (DSL invariant: must start with
``Flux``). Iteration order is the sorted alphabetic order of the spec
files on disk — this keeps ``catalog.json`` + emitted factories
deterministic without relying on the filesystem's natural order.
"""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
from pathlib import Path
from typing import Any


def _loader_script(root: Path) -> Path:
    return root / "shared" / "scripts" / "flux" / "_loader.ts"


def _ts_workspace(root: Path) -> Path:
    return root / "ts"


def _spec_files(root: Path) -> list[Path]:
    specs_dir = root / "catalog" / "flux" / "specs"
    if not specs_dir.is_dir():
        return []
    return sorted(p for p in specs_dir.glob("*.spec.ts") if p.is_file())


def _run_loader_on(root: Path, spec_path: Path) -> dict[str, Any]:
    ts_dir = _ts_workspace(root)
    if not (ts_dir / "node_modules").is_dir():
        raise RuntimeError(f"flux loader: ts/ workspace missing node_modules — run `cd {ts_dir} && npm install` first")
    if shutil.which("npx") is None:
        raise RuntimeError("flux loader: `npx` not found on PATH (install Node.js 20+)")
    env = os.environ.copy()
    env["NODE_PATH"] = str((ts_dir / "node_modules").resolve())
    cmd = [
        "npx",
        "--yes",
        "tsx",
        str(_loader_script(root).resolve()),
        str(spec_path.resolve()),
    ]
    proc = subprocess.run(
        cmd,
        cwd=str(ts_dir),
        env=env,
        check=False,
        capture_output=True,
        text=True,
    )
    if proc.returncode != 0:
        sys.stderr.write(proc.stderr)
        raise RuntimeError(f"flux loader: _loader.ts failed for {spec_path.name} (exit {proc.returncode})")
    try:
        return json.loads(proc.stdout)
    except json.JSONDecodeError as exc:
        sys.stderr.write(proc.stderr)
        raise RuntimeError(f"flux loader: _loader.ts produced non-JSON output for {spec_path.name}: {exc}") from exc


def load_all(root: Path) -> dict[str, dict[str, Any]]:
    """Load every spec under ``catalog/flux/specs/``, keyed by component name."""
    specs: dict[str, dict[str, Any]] = {}
    for spec_path in _spec_files(root):
        data = _run_loader_on(root, spec_path)
        name = data.get("name")
        if not isinstance(name, str):
            raise RuntimeError(f"flux loader: {spec_path.name} emitted no .name field")
        if name in specs:
            raise RuntimeError(f"flux loader: duplicate component name {name!r} across spec files")
        specs[name] = data
    return specs


def load_one(root: Path, spec_path: Path) -> dict[str, Any]:
    """Load a single spec file. Primarily a test hook."""
    return _run_loader_on(root, spec_path)
